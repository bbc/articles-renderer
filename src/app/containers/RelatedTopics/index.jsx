import React, { useContext } from 'react';
import { TopicTag, TopicTags } from '@bbc/psammead-topic-tags';
import { pathOr } from 'ramda';
import SectionLabel from '@bbc/psammead-section-label';
import styled from '@emotion/styled';
import { arrayOf, shape, string } from 'prop-types';
import { GEL_SPACING_QUIN } from '@bbc/gel-foundations/spacings';
import { ServiceContext } from '#app/contexts/ServiceContext';
import { RequestContext } from '#app/contexts/RequestContext';
import useClickTrackerHandler from '#hooks/useClickTrackerHandler';
import useViewTracker from '#hooks/useViewTracker';

const eventTrackingData = {
  componentName: 'topics',
};

const StyledTopicsWrapper = styled.aside`
  padding-bottom: ${GEL_SPACING_QUIN};
`;

const StyledSectionLabel = styled(SectionLabel)`
  margin-top: 0;
`;

const RelatedTopics = ({ topics }) => {
  const { service, script, translations, dir } = useContext(ServiceContext);
  const { variant } = useContext(RequestContext);
  const isEnabled =
    process?.env?.SIMORGH_APP_ENV === 'test' ||
    process?.env?.SIMORGH_APP_ENV === 'local';
  const clickTrackerHandler = useClickTrackerHandler(eventTrackingData);
  const viewRef = useViewTracker(eventTrackingData);
  const heading = pathOr('Related Topics', ['relatedTopics'], translations);

  const getTopicPageUrl = id => {
    return variant
      ? `${process?.env?.SIMORGH_BASE_URL}/${service}/${variant}/topics/${id}`
      : `${process?.env?.SIMORGH_BASE_URL}/${service}/topics/${id}`;
  };

  return (
    topics &&
    isEnabled && (
      <StyledTopicsWrapper
        aria-labelledby="related-topics"
        role="complementary"
      >
        <StyledSectionLabel
          bar
          script={script}
          service={service}
          dir={dir}
          labelId="related-topics"
        >
          {heading}
        </StyledSectionLabel>
        <TopicTags service={service} script={script}>
          {topics.length === 1 ? (
            <TopicTag
              name={topics[0].topicName}
              link={getTopicPageUrl(topics[0].topicId)}
              onClick={clickTrackerHandler}
              ref={viewRef}
              key={topics[0].topicId}
            />
          ) : (
            topics.map(({ topicName, topicId }) => (
              <TopicTag
                name={topicName}
                link={getTopicPageUrl(topicId)}
                onClick={clickTrackerHandler}
                ref={viewRef}
                key={topicId}
              />
            ))
          )}
        </TopicTags>
      </StyledTopicsWrapper>
    )
  );
};

RelatedTopics.propTypes = {
  topics: arrayOf(
    shape({
      topicName: string,
      topicId: string,
    }),
  ),
};

RelatedTopics.defaultProps = {
  topics: null,
};

export default RelatedTopics;
