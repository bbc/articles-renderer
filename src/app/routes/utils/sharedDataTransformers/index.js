import augmentWithTimestamp from './augmentWithTimestamp';
import applyBlockPositioning from './applyBlockPositioning';
import addIdsToBlocks from './addIdsToBlocks';
import addIndexesToEmbeds from './addIndexesToEmbeds';

export {
  augmentWithTimestamp,
  applyBlockPositioning,
  addIdsToBlocks,
  addIndexesToEmbeds,
};
