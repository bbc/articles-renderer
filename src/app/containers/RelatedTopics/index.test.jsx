/* eslint-disable react/prop-types */
import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { render } from '@testing-library/react';
import { RequestContextProvider } from '#contexts/RequestContext';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { ToggleContextProvider } from '#contexts/ToggleContext';
import { EventTrackingContextProvider } from '#contexts/EventTrackingContext';
import RelatedTopics from '#containers/RelatedTopics';
import * as clickTracker from '#hooks/useClickTrackerHandler';
import * as viewTracker from '#hooks/useViewTracker';
import { STORY_PAGE, ARTICLE_PAGE } from '#app/routes/utils/pageTypes';

process.env.SIMORGH_BASE_URL = 'https://bbc.com';
const { env } = process;

beforeEach(() => {
  jest.resetModules();
  process.env.SIMORGH_APP_ENV = 'test';
});

const WithContexts = ({
  children,
  variant,
  service = 'mundo',
  isAmp = false,
  pageType = STORY_PAGE,
}) => {
  return (
    <RequestContextProvider
      service={service}
      variant={variant}
      pageType={pageType}
      isAmp={isAmp}
      pathname="/"
    >
      <ServiceContextProvider service={service}>
        <ToggleContextProvider
          toggles={{
            eventTracking: {
              enabled: true,
            },
          }}
        >
          <EventTrackingContextProvider>
            {children}
          </EventTrackingContextProvider>
        </ToggleContextProvider>
      </ServiceContextProvider>
    </RequestContextProvider>
  );
};

describe('Expected use', () => {
  afterAll(() => {
    process.env = env;
  });

  shouldMatchSnapshot(
    'should render correctly with no tags',
    <WithContexts>
      <RelatedTopics />
    </WithContexts>,
  );

  shouldMatchSnapshot(
    'should render correctly with a single tag',
    <WithContexts>
      <RelatedTopics topics={[{ topicName: 'topic', topicId: '123' }]} />
    </WithContexts>,
  );

  shouldMatchSnapshot(
    'should render correctly with multiple tags',
    <WithContexts>
      <RelatedTopics
        topics={[
          { topicName: 'topic1', topicId: '1' },
          { topicName: 'topic2', topicId: '2' },
          { topicName: 'topic3', topicId: '3' },
        ]}
      />
      ,
    </WithContexts>,
  );

  it('should construct the correct topics href given a topic id without a variant', () => {
    const topic = {
      topicName: 'foo',
      topicId: 'bar',
    };

    const { getByText } = render(
      <WithContexts service="pidgin">
        <RelatedTopics topics={[topic]} />
      </WithContexts>,
    );

    expect(getByText(topic.topicName)).toHaveAttribute(
      'href',
      `https://bbc.com/pidgin/topics/${topic.topicId}`,
    );
  });

  it('should construct the correct topics href given a topic id and a variant', () => {
    const topic = {
      topicName: 'foo',
      topicId: 'bar',
    };

    const { getByText } = render(
      <WithContexts service="uzbek" variant="cyr">
        <RelatedTopics topics={[topic]} />
      </WithContexts>,
    );

    expect(getByText(topic.topicName)).toHaveAttribute(
      'href',
      `https://bbc.com/uzbek/cyr/topics/${topic.topicId}`,
    );
  });

  it('should return null when on the live environment', () => {
    process.env.SIMORGH_APP_ENV = 'live';
    const { container } = render(
      <WithContexts optimoEnabled={false} pageType={ARTICLE_PAGE}>
        <RelatedTopics
          topics={[
            { topicName: 'topic1', topicId: '1' },
            { topicName: 'topic2', topicId: '2' },
            { topicName: 'topic3', topicId: '3' },
          ]}
        />
      </WithContexts>,
    );
    expect(container.firstChild).toBeNull();
  });

  it('should return null when process is undefined', () => {
    process.env = undefined;

    const { container } = render(
      <WithContexts optimoEnabled={false} pageType={ARTICLE_PAGE}>
        <RelatedTopics
          topics={[
            { topicName: 'topic1', topicId: '1' },
            { topicName: 'topic2', topicId: '2' },
            { topicName: 'topic3', topicId: '3' },
          ]}
        />
      </WithContexts>,
    );
    expect(container.firstChild).toBeNull();
  });
});

describe('A11y', () => {
  it('should not render an unordered list when there is only one topic', () => {
    const { container } = render(
      <WithContexts>
        <RelatedTopics topics={[{ topicName: 'topic1', topicId: '1' }]} />
      </WithContexts>,
    );
    expect(container.querySelector('ul')).toBeNull();
  });

  it('should render an unordered list when there is more than one topic', () => {
    const { container } = render(
      <WithContexts>
        <RelatedTopics
          topics={[
            { topicName: 'topic1', topicId: '1' },
            { topicName: 'topic2', topicId: '2' },
          ]}
        />
      </WithContexts>,
    );
    expect(container.querySelector('ul')).not.toBeNull();
  });
});

describe('Event Tracking', () => {
  const eventTrackingData = {
    componentName: 'topics',
  };

  it('should call the click tracker with the correct params', () => {
    const clickTrackerSpy = jest.spyOn(clickTracker, 'default');
    render(
      <WithContexts>
        <RelatedTopics topics={[{ topicName: 'topic', topicId: 'id' }]} />
      </WithContexts>,
    );

    expect(clickTrackerSpy).toHaveBeenCalledWith(eventTrackingData);
  });

  it('should call the view tracker with the correct params', () => {
    const viewTrackerSpy = jest.spyOn(viewTracker, 'default');
    render(
      <WithContexts>
        <RelatedTopics topics={[{ topicName: 'topic', topicId: 'id' }]} />
      </WithContexts>,
    );

    expect(viewTrackerSpy).toHaveBeenCalledWith(eventTrackingData);
  });
});
