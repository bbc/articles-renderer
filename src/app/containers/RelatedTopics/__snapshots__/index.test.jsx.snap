// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expected use should render correctly with a single tag 1`] = `
.emotion-0 {
  padding-bottom: 2.5rem;
}

.emotion-3 {
  position: relative;
  z-index: 0;
  color: #3F3F42;
  margin-top: 2rem;
  margin-top: 0;
}

@media (min-width: 37.5rem) {
  .emotion-3 {
    margin-top: 1.5rem;
  }
}

@media (min-width: 63rem) {
  .emotion-3 {
    margin-bottom: 1.5rem;
  }
}

.emotion-5 {
  border-top: 0.0625rem solid #AEAEB5;
  z-index: -1;
}

@media screen and (-ms-high-contrast: active) {
  .emotion-5 {
    border-color: windowText;
  }
}

@media (min-width: 37.5rem) {
  .emotion-5 {
    position: absolute;
    left: 0;
    right: 0;
    top: 1.5rem;
  }
}

.emotion-7 {
  margin: 0;
  padding: 0;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row nowrap;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-height: 2.75rem;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

@media (min-width: 37.5rem) {
  .emotion-11 {
    -webkit-align-items: stretch;
    -webkit-box-align: stretch;
    -ms-flex-align: stretch;
    align-items: stretch;
  }
}

.emotion-13 {
  font-size: 1.125rem;
  line-height: 1.5rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  background-color: #FDFDFD;
  margin: 1rem 0;
  padding-right: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-13 {
    font-size: 1.25rem;
    line-height: 1.625rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-13 {
    font-size: 1.5rem;
    line-height: 2rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-13 {
    margin: 0;
  }
}

@media (min-width: 37.5rem) {
  .emotion-13 {
    padding-right: 1rem;
  }
}

.emotion-15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-top: -1rem;
  margin-bottom: 0;
  margin-left: -0.25rem;
  margin-right: -0.25rem;
  padding: 0;
}

.emotion-17 {
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.875rem;
  line-height: 1.125rem;
  word-break: break-word;
  margin-top: 1rem;
  margin-left: 0.25rem;
  margin-right: 0.25rem;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-17 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-17 {
    font-size: 0.8125rem;
    line-height: 1rem;
  }
}

.emotion-17 a {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0.96875rem 0.4375rem;
  line-height: 1;
  background-color: #F2F2F2;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #222222;
}

.emotion-17 a:hover,
.emotion-17 a:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-17 a:visited {
  color: #6E6E73;
}

<div>
  <aside
    aria-labelledby="related-topics"
    class="emotion-0 emotion-1"
    role="complementary"
  >
    <div
      class="emotion-2 emotion-3 emotion-4"
    >
      <div
        class="emotion-5 emotion-6"
      />
      <h2
        class="emotion-7 emotion-8"
      >
        <span
          class="emotion-9 emotion-10"
        >
          <span
            class="emotion-11 emotion-12"
          >
            <span
              class="emotion-13 emotion-14"
              dir="ltr"
              id="related-topics"
            >
              Temas relacionados
            </span>
          </span>
        </span>
      </h2>
    </div>
    <div
      class="emotion-15 emotion-16"
    >
      <div
        class="emotion-17 emotion-18"
      >
        <a
          href="https://bbc.com/mundo/topics/123"
        >
          topic
        </a>
      </div>
    </div>
  </aside>
</div>
`;

exports[`Expected use should render correctly with multiple tags 1`] = `
.emotion-0 {
  padding-bottom: 2.5rem;
}

.emotion-3 {
  position: relative;
  z-index: 0;
  color: #3F3F42;
  margin-top: 2rem;
  margin-top: 0;
}

@media (min-width: 37.5rem) {
  .emotion-3 {
    margin-top: 1.5rem;
  }
}

@media (min-width: 63rem) {
  .emotion-3 {
    margin-bottom: 1.5rem;
  }
}

.emotion-5 {
  border-top: 0.0625rem solid #AEAEB5;
  z-index: -1;
}

@media screen and (-ms-high-contrast: active) {
  .emotion-5 {
    border-color: windowText;
  }
}

@media (min-width: 37.5rem) {
  .emotion-5 {
    position: absolute;
    left: 0;
    right: 0;
    top: 1.5rem;
  }
}

.emotion-7 {
  margin: 0;
  padding: 0;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row nowrap;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-height: 2.75rem;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

@media (min-width: 37.5rem) {
  .emotion-11 {
    -webkit-align-items: stretch;
    -webkit-box-align: stretch;
    -ms-flex-align: stretch;
    align-items: stretch;
  }
}

.emotion-13 {
  font-size: 1.125rem;
  line-height: 1.5rem;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  background-color: #FDFDFD;
  margin: 1rem 0;
  padding-right: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-13 {
    font-size: 1.25rem;
    line-height: 1.625rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-13 {
    font-size: 1.5rem;
    line-height: 2rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-13 {
    margin: 0;
  }
}

@media (min-width: 37.5rem) {
  .emotion-13 {
    padding-right: 1rem;
  }
}

.emotion-15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-top: -1rem;
  margin-bottom: 0;
  margin-left: -0.25rem;
  margin-right: -0.25rem;
  padding: 0;
  list-style-type: none;
}

.emotion-17 {
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.875rem;
  line-height: 1.125rem;
  word-break: break-word;
  margin-top: 1rem;
  margin-left: 0.25rem;
  margin-right: 0.25rem;
}

@media (min-width: 20rem) and (max-width: 37.4375rem) {
  .emotion-17 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

@media (min-width: 37.5rem) {
  .emotion-17 {
    font-size: 0.8125rem;
    line-height: 1rem;
  }
}

.emotion-17 a {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0.96875rem 0.4375rem;
  line-height: 1;
  background-color: #F2F2F2;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #222222;
}

.emotion-17 a:hover,
.emotion-17 a:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-17 a:visited {
  color: #6E6E73;
}

<div>
  <aside
    aria-labelledby="related-topics"
    class="emotion-0 emotion-1"
    role="complementary"
  >
    <div
      class="emotion-2 emotion-3 emotion-4"
    >
      <div
        class="emotion-5 emotion-6"
      />
      <h2
        class="emotion-7 emotion-8"
      >
        <span
          class="emotion-9 emotion-10"
        >
          <span
            class="emotion-11 emotion-12"
          >
            <span
              class="emotion-13 emotion-14"
              dir="ltr"
              id="related-topics"
            >
              Temas relacionados
            </span>
          </span>
        </span>
      </h2>
    </div>
    <ul
      class="emotion-15 emotion-16"
      role="list"
    >
      <li
        class="emotion-17 emotion-18"
      >
        <a
          href="https://bbc.com/mundo/topics/1"
        >
          topic1
        </a>
      </li>
      <li
        class="emotion-17 emotion-18"
      >
        <a
          href="https://bbc.com/mundo/topics/2"
        >
          topic2
        </a>
      </li>
      <li
        class="emotion-17 emotion-18"
      >
        <a
          href="https://bbc.com/mundo/topics/3"
        >
          topic3
        </a>
      </li>
    </ul>
  </aside>
  ,
</div>
`;

exports[`Expected use should render correctly with no tags 1`] = `<div />`;
